File: sdk/es/test/validate.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import {
  validateRequest,
  ProtocolError,
  entry,
  type AuthorizationRequest,
} from "../src/stepauth/index.js";

// Timestamps are relative to now: validation checks skew against the local
// clock, so a fixed date would start rejecting five minutes after it was written.
function baseRequest(): AuthorizationRequest {
  const now = new Date();
  return {
    requestId: "req_1",
    senderId: "sp.example.com",
    recipientId: "hub.example.com",
    timestamp: now.toISOString(),
    expiresAt: new Date(now.getTime() + 30 * 60_000).toISOString(),
    principal: { subject: "email:a@b.com" },
    action: { type: "acme.widget.delete", category: "data.delete", summary: "Delete widget" },
  };
}

function codeOf(fn: () => void): string {
  try {
    fn();
  } catch (err) {
    assert.ok(err instanceof ProtocolError);
    return err.code;
  }
  throw new Error("expected validateRequest to throw");
}

test("validateRequest accepts a well-formed request", () => {
  assert.doesNotThrow(() => validateRequest(baseRequest()));
});

test("validateRequest rejects a missing required field", () => {
  const req = baseRequest();
  req.requestId = "";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts requestId at the 256-byte bound", () => {
  const req = baseRequest();
  req.requestId = "r".repeat(256);
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects requestId over the 256-byte bound", () => {
  const req = baseRequest();
  req.requestId = "r".repeat(257);
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest rejects a requestId containing '#'", () => {
  const req = baseRequest();
  req.requestId = "req#1";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest rejects a senderId with a non-printable byte", () => {
  const req = baseRequest();
  req.senderId = "sp\t.example.com";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts action.name at the 120-codepoint bound", () => {
  const req = baseRequest();
  req.action.name = "a".repeat(120);
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects action.name over 120 codepoints", () => {
  const req = baseRequest();
  req.action.name = "a".repeat(121);
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

// The divergence this layer is most likely to introduce: JS string length
// counts UTF-16 code units, so an emoji (astral, 2 code units) would be
// double-counted against a rune-based limit unless action.name is measured by
// codepoint, matching Go's utf8.RuneCountInString.
test("validateRequest counts action.name by codepoint, not UTF-16 code unit", () => {
  const req = baseRequest();
  // 119 ASCII chars + one astral emoji = 120 codepoints, but 121 UTF-16 units.
  req.action.name = "a".repeat(119) + "\u{1F600}";
  assert.equal(req.action.name.length, 121);
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects action.name over 120 codepoints even with astral characters", () => {
  const req = baseRequest();
  req.action.name = "a".repeat(120) + "\u{1F600}";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest rejects an unknown action.category", () => {
  const req = baseRequest();
  // @ts-expect-error deliberately outside the closed registry
  req.action.category = "data.frobnicate";
  assert.equal(codeOf(() => validateRequest(req)), "unknown_category");
});

test("validateRequest accepts every registered category", () => {
  const req = baseRequest();
  req.action.category = "financial.approve";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects a principal.subject that is not an individual NameID", () => {
  const req = baseRequest();
  req.principal.subject = "group:eng";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts a persistent principal.subject", () => {
  const req = baseRequest();
  req.principal.subject = "persistent:p_123";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects a malformed action.approver", () => {
  const req = baseRequest();
  req.action.approver = "not-a-nameid";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts a well-formed action.approver", () => {
  const req = baseRequest();
  req.action.approver = "group:eng";
  assert.doesNotThrow(() => validateRequest(req));
});

// The hub falls back to the app's registered callback when callbackUrl is
// omitted; an empty string must hit that same fallback, not invalid_callback_url.
test("validateRequest accepts an empty callbackUrl", () => {
  const req = baseRequest();
  req.callbackUrl = "";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest accepts an empty action.approver", () => {
  const req = baseRequest();
  req.action.approver = "";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects an operator chain deeper than 5", () => {
  const req = baseRequest();
  let op: AuthorizationRequest["principal"]["operator"] = { subject: "email:l6@b.com" };
  for (let i = 0; i < 5; i++) op = { subject: `email:l${i}@b.com`, operator: op };
  req.principal.operator = op;
  assert.equal(codeOf(() => validateRequest(req)), "operator_depth_exceeded");
});

test("validateRequest accepts an operator chain exactly 5 deep", () => {
  const req = baseRequest();
  let op: AuthorizationRequest["principal"]["operator"] = { subject: "email:l5@b.com" };
  for (let i = 0; i < 4; i++) op = { subject: `email:l${i}@b.com`, operator: op };
  req.principal.operator = op;
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest accepts an https callbackUrl", () => {
  const req = baseRequest();
  req.callbackUrl = "https://sp.example.com/hooks/stepauth";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest accepts an http callbackUrl to localhost", () => {
  const req = baseRequest();
  req.callbackUrl = "http://localhost:4000/hooks/stepauth";
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects a non-loopback http callbackUrl", () => {
  const req = baseRequest();
  req.callbackUrl = "http://sp.example.com/hooks/stepauth";
  assert.equal(codeOf(() => validateRequest(req)), "invalid_callback_url");
});

test("validateRequest rejects an unparseable callbackUrl", () => {
  const req = baseRequest();
  req.callbackUrl = "not a url";
  assert.equal(codeOf(() => validateRequest(req)), "invalid_callback_url");
});

test("validateRequest rejects a duplicate labeled-entry key in action.details", () => {
  const req = baseRequest();
  req.action.details = [entry("amount", "Amount", "1"), entry("amount", "Amount", "2")];
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts unique labeled-entry keys in principal.attributes", () => {
  const req = baseRequest();
  req.principal.attributes = [entry("department", "Department", "eng")];
  assert.doesNotThrow(() => validateRequest(req));
});

test("validateRequest rejects a non-RFC3339 timestamp", () => {
  const req = baseRequest();
  req.timestamp = "not-a-time";
  assert.equal(codeOf(() => validateRequest(req)), "timestamp_out_of_range");
});

test("validateRequest rejects a non-RFC3339 expiresAt", () => {
  const req = baseRequest();
  req.expiresAt = "not-a-time";
  assert.equal(codeOf(() => validateRequest(req)), "malformed_request");
});

test("validateRequest accepts RFC3339 timestamps with an explicit offset", () => {
  const req = baseRequest();
  // The same instant as now, written with a -04:00 offset instead of Z, so this
  // exercises offset parsing without tripping the skew check.
  const now = new Date();
  const local = new Date(now.getTime() - 4 * 60 * 60_000).toISOString().replace(/Z$/, "-04:00");
  req.timestamp = local;
  req.expiresAt = new Date(now.getTime() + 30 * 60_000).toISOString();
  assert.doesNotThrow(() => validateRequest(req));
});

File: sdk/go/stepauth/validate_test.go
package stepauth

import (
	"errors"
	"strings"
	"testing"
)

// protocolCode recovers the wire code from a Validate error, which is an
// error interface rather than a concrete type.
func protocolCode(t *testing.T, err error) Code {
	t.Helper()
	var pe *ProtocolError
	if !errors.As(err, &pe) {
		t.Fatalf("error is not a *ProtocolError: %v", err)
	}
	return pe.Code
}

func TestValidateAcceptsWellFormedRequest(t *testing.T) {
	if err := Validate(validRequest()); err != nil {
		t.Errorf("Validate() = %v, want nil", err)
	}
}

func TestValidateRejectsMissingRequiredFields(t *testing.T) {
	tests := []struct {
		name   string
		mutate func(*AuthorizationRequest)
	}{
		{"requestId", func(r *AuthorizationRequest) { r.RequestID = "" }},
		{"senderId", func(r *AuthorizationRequest) { r.SenderID = "" }},
		{"recipientId", func(r *AuthorizationRequest) { r.RecipientID = "" }},
		{"timestamp", func(r *AuthorizationRequest) { r.Timestamp = "" }},
		{"expiresAt", func(r *AuthorizationRequest) { r.ExpiresAt = "" }},
		{"action.type", func(r *AuthorizationRequest) { r.Action.Type = "" }},
		{"action.category", func(r *AuthorizationRequest) { r.Action.Category = "" }},
		{"action.summary", func(r *AuthorizationRequest) { r.Action.Summary = "" }},
		{"principal.subject", func(r *AuthorizationRequest) { r.Principal.Subject = "" }},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := validRequest()
			tt.mutate(req)
			if err := Validate(req); err == nil {
				t.Errorf("Validate() accepted a request missing %s", tt.name)
			}
		})
	}
}

func TestValidateRejectsKeyIDWithHash(t *testing.T) {
	req := validRequest()
	req.RequestID = "req#1"
	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted a requestId containing '#'")
	}
	if code := protocolCode(t, err); code != CodeMalformedRequest {
		t.Errorf("Code = %q, want %q", code, CodeMalformedRequest)
	}
}

func TestValidateRejectsOverlongKeyID(t *testing.T) {
	req := validRequest()
	req.SenderID = strings.Repeat("a", MaxKeyIDLen+1)
	if Validate(req) == nil {
		t.Error("Validate() accepted a senderId over MaxKeyIDLen")
	}
}

func TestValidateAcceptsMaxLengthKeyID(t *testing.T) {
	req := validRequest()
	req.SenderID = strings.Repeat("a", MaxKeyIDLen)
	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for a MaxKeyIDLen senderId", err)
	}
}

func TestValidateRejectsNonPrintableKeyID(t *testing.T) {
	req := validRequest()
	req.RequestID = "req\x01id"
	if Validate(req) == nil {
		t.Error("Validate() accepted a requestId with a non-printable byte")
	}
}

// TestValidateActionNameCountsRunesNotBytes is the known cross-language
// divergence risk: a byte-length check would reject a multi-byte name well
// under the 120-rune limit.
func TestValidateActionNameCountsRunesNotBytes(t *testing.T) {
	req := validRequest()
	req.Action.Name = strings.Repeat("é", 120) // 120 runes, 240 bytes

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for a 120-rune multi-byte name", err)
	}
}

func TestValidateRejectsActionNameOver120Runes(t *testing.T) {
	req := validRequest()
	req.Action.Name = strings.Repeat("é", 121)

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted a 121-rune action.name")
	}
	if code := protocolCode(t, err); code != CodeMalformedRequest {
		t.Errorf("Code = %q, want %q", code, CodeMalformedRequest)
	}
}

func TestValidateRejectsUnknownCategory(t *testing.T) {
	req := validRequest()
	req.Action.Category = "not.a.real.category"

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted an unregistered category")
	}
	if code := protocolCode(t, err); code != CodeUnknownCategory {
		t.Errorf("Code = %q, want %q", code, CodeUnknownCategory)
	}
}

func TestValidateAcceptsEveryRegisteredCategory(t *testing.T) {
	for c := range categoryRegistry {
		req := validRequest()
		req.Action.Category = c
		if err := Validate(req); err != nil {
			t.Errorf("Validate() = %v, want nil for category %q", err, c)
		}
	}
}

func TestValidateRejectsNonIndividualSubject(t *testing.T) {
	req := validRequest()
	req.Principal.Subject = GroupID("g_1")

	if Validate(req) == nil {
		t.Error("Validate() accepted a group NameID as principal.subject")
	}
}

func TestValidateAcceptsPersistentSubject(t *testing.T) {
	req := validRequest()
	req.Principal.Subject = Persistent("p_1")

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for a persistent subject", err)
	}
}

func TestValidateRejectsMalformedApprover(t *testing.T) {
	req := validRequest()
	req.Action.Approver = "not-a-nameid"

	if Validate(req) == nil {
		t.Error("Validate() accepted a malformed action.approver")
	}
}

func TestValidateAcceptsWellFormedApprover(t *testing.T) {
	req := validRequest()
	req.Action.Approver = GroupID("g_reviewers")

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for a well-formed group approver", err)
	}
}

func TestValidateRejectsOperatorDepthOver5(t *testing.T) {
	req := validRequest()
	p := &req.Principal
	for range 6 {
		p.Operator = &Principal{Subject: Email("op@b.com")}
		p = p.Operator
	}

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted an operator chain deeper than 5")
	}
	if code := protocolCode(t, err); code != CodeOperatorDepthExceeded {
		t.Errorf("Code = %q, want %q", code, CodeOperatorDepthExceeded)
	}
}

func TestValidateAcceptsOperatorDepthOf5(t *testing.T) {
	req := validRequest()
	p := &req.Principal
	for range 5 {
		p.Operator = &Principal{Subject: Email("op@b.com")}
		p = p.Operator
	}

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for an operator chain of depth 5", err)
	}
}

func TestValidateRejectsBadTimestamp(t *testing.T) {
	req := validRequest()
	req.Timestamp = "not-a-timestamp"

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted a non-RFC3339 timestamp")
	}
	if code := protocolCode(t, err); code != CodeTimestampOutOfRange {
		t.Errorf("Code = %q, want %q", code, CodeTimestampOutOfRange)
	}
}

func TestValidateRejectsBadExpiresAt(t *testing.T) {
	req := validRequest()
	req.ExpiresAt = "not-a-timestamp"

	if Validate(req) == nil {
		t.Error("Validate() accepted a non-RFC3339 expiresAt")
	}
}

func TestValidateRejectsClockSkewOver5Minutes(t *testing.T) {
	req := validRequest()
	req.Timestamp = "2020-01-01T00:00:00Z" // far in the past relative to time.Now

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted a timestamp with excessive skew")
	}
	if code := protocolCode(t, err); code != CodeTimestampOutOfRange {
		t.Errorf("Code = %q, want %q", code, CodeTimestampOutOfRange)
	}
}

func TestValidateCallbackURLAcceptsHTTPS(t *testing.T) {
	req := validRequest()
	req.CallbackURL = "https://sp.example.com/callback"

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for an https callbackUrl", err)
	}
}

func TestValidateCallbackURLAcceptsLoopback(t *testing.T) {
	for _, u := range []string{"http://localhost/callback", "http://127.0.0.1:8080/callback"} {
		req := validRequest()
		req.CallbackURL = u
		if err := Validate(req); err != nil {
			t.Errorf("Validate() = %v, want nil for loopback callbackUrl %q", err, u)
		}
	}
}

func TestValidateCallbackURLRejectsPlainHTTP(t *testing.T) {
	req := validRequest()
	req.CallbackURL = "http://sp.example.com/callback"

	err := Validate(req)
	if err == nil {
		t.Fatal("Validate() accepted a non-loopback http callbackUrl")
	}
	if code := protocolCode(t, err); code != CodeInvalidCallbackURL {
		t.Errorf("Code = %q, want %q", code, CodeInvalidCallbackURL)
	}
}

func TestValidateAllowsEmptyCallbackURL(t *testing.T) {
	req := validRequest()
	req.CallbackURL = ""

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for an omitted callbackUrl", err)
	}
}

func TestValidateRejectsDuplicateAttributeKeys(t *testing.T) {
	req := validRequest()
	req.Principal.Attributes = Entries(Entry("dup", "A", "1"), Entry("dup", "B", "2"))

	if Validate(req) == nil {
		t.Error("Validate() accepted duplicate principal.attributes keys")
	}
}

func TestValidateAcceptsDistinctAttributeKeys(t *testing.T) {
	req := validRequest()
	req.Principal.Attributes = Entries(Entry("dept", "Department", "eng"), Entry("role", "Role", "admin"))

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for distinct attribute keys", err)
	}
}

func TestValidateAcceptsCamelCaseDetailKey(t *testing.T) {
	req := validRequest()
	req.Action.Details = Entries(Entry("userId", "User ID", "1"))

	if err := Validate(req); err != nil {
		t.Errorf("Validate() = %v, want nil for a camelCase action.details key", err)
	}
}
